pkg/virt-launcher/virtwrap/agent-poller/agent_poller_test.go

```diff
@@ -20,413 +20,140 @@
 
-			users, err := parseUsers(jsonInput)
-			expectedUsers := []api.User{
-				api.User{
-					Name:      "bob",
-					Domain:    "bobs",
-					LoginTime: 99999,
-				},
-			}
-
-			Expect(err).ToNot(HaveOccurred(), "users should be parsed normally")
-			Expect(users).To(Equal(expectedUsers))
+			Expect(commandExecutions).To(Equal(expectedExecutions))
 		})
 
-		Context("with AsyncAgentStore", func() {
-
-			It("should store and load the data", func() {
-				var agentStore = NewAsyncAgentStore()
-				agentVersion := "4.1"
-				agentStore.Store(GET_AGENT, agentVersion)
-				agent := agentStore.GetGA()
+		It("executes the agent commands based on the time interval specified", func() {
+			const interval = 1
+			const expectedExecutions = 3
 
-				Expect(agent).To(Equal(agentVersion))
-			})
+			commandExecutions := runPollAndCountCommandExecution(interval, expectedExecutions, 0)
 
-			It("should fire an event for new sysinfo data", func() {
-				var agentStore = NewAsyncAgentStore()
-
-				fakeInfo := api.GuestOSInfo{
-					Name:          "TestGuestOSName",
-					KernelRelease: "1.1.0-Generic",
-					Version:       "1.0.0",
-					PrettyName:    "TestGuestOSName 1.0.0",
-					VersionId:     "1.0.0",
-					KernelVersion: "1.1.0",
-					Machine:       "x86_64",
-					Id:            "testguestos",
-				}
-				agentStore.Store(GET_OSINFO, fakeInfo)
-
-				Expect(agentStore.AgentUpdated).To(Receive(Equal(AgentUpdatedEvent{
-					Type:       GET_OSINFO,
-					DomainInfo: api.DomainGuestInfo{OSInfo: &fakeInfo},
-				})))
-			})
+			Expect(commandExecutions).To(Equal(expectedExecutions))
+		})
 
-			It("should not fire an event for the same sysinfo data", func() {
-				var agentStore = NewAsyncAgentStore()
-				fakeInfo := api.GuestOSInfo{
-					Name:          "TestGuestOSName",
-					KernelRelease: "1.1.0-Generic",
-					Version:       "1.0.0",
-					PrettyName:    "TestGuestOSName 1.0.0",
-					VersionId:     "1.0.0",
-					KernelVersion: "1.1.0",
-					Machine:       "x86_64",
-					Id:            "testguestos",
-				}
+		It("executes the agent commands in increasing intervals at initial run", func() {
+			const interval = 30
+			const expectedExecutions = 2
 
-				agentStore.Store(GET_OSINFO, fakeInfo)
-				Expect(agentStore.AgentUpdated).To(Receive(Equal(AgentUpdatedEvent{
-					Type:       GET_OSINFO,
-					DomainInfo: api.DomainGuestInfo{OSInfo: &fakeInfo},
-				})))
+			t := pollInitialInterval + timeoutBuffer
+			commandExecutions := runPollAndCountCommandExecution(interval, expectedExecutions, t)
 
-				agentStore.Store(GET_OSINFO, fakeInfo)
-				Expect(agentStore.AgentUpdated).ToNot(Receive())
-			})
+			Expect(commandExecutions).To(Equal(expectedExecutions))
 		})
 	})
 })
+
+// countSignals counts the number of signals received through the `done` channel
+// The timeout is calculated from the first received signal.
+func countSignals(done <-chan struct{}, timeout time.Duration) int {
+	var counter int
+	t := time.After(timeout)
```

Comment:
Dont you want to remove this ?
according your last push, start the timer after the first counter has received

Reply:
@oshoval In the current approach, the timeout is being  implicitly increased. The amount that was implicitly added is equal to the delay of the launch of the poll worker go routine. This amount can also be considered as a toleration to a possible drifting of the timer ticks.

Reply:
i compare this line and the added line here
https://github.com/kubevirt/kubevirt/compare/b42fcf0bc785de50d1c2a658a7ac42f58eadf8a0..7f975efb6833d6a5cd6902f6da4ddfefeb7f52d0

else i dont understand the comment here
https://github.com/kubevirt/kubevirt/pull/4512#issuecomment-727613924
Since the timer on line 147 might stop the loop and the test will fail, before we counted one element.

Reply:
@oshoval I think that the word "start" should be replaced with "restart". As per my understanding, the timer is first set before the select, allowing the go routine to start with a delay of the timeout value. And then, after the first message is received, the timer has been restarted again.

Reply:
https://github.com/kubevirt/kubevirt/pull/4512/files#diff-9782b02222c8cc340e9fbb694311494f4fd32c70e4519596f8e64736a92a84f0R103
I think this is the go routine we need to avoid race between this timer case<-t and the go routine.

Reply:
The fist timer is set in order to treat cases where there is no signal coming at all. The second is to restart the timer so it will be in sync with the producer for cases where it took it too long to launch.

Which part is not clear or where do you see the problem?

Reply:
The assumption (even if would work on 99.9% of the time) that the scheduler would schedule the go w.Poll in ~10msec,
before the timer expires.

Reply:
Where do you see this assumption? Walk the code and show me please.

Reply:
In case `t == 0`, and `expectedIterations = 1` , the timeout would be the 10msec isnt it?
I didnt recheck if we still have the above tuple in the tests, but in one of the previous reviews we had i think,
anyway, the function should be agnostic of its usages i think ?

Reply:
So all this comment thread is about one specific case where the inputted timeout value is the minimum, i.e. the drift of 10msec. In such a case, we wait only 10msec and you claim the goroutine may still not have started.

So please tell me, how much time it should take a go-routine to start?

Reply:
I look on it otherwise,
Even if it takes 700ns 99.9% of the time and one time it will take 10ms because HD is full and there is lots of thrashing and page faults, i would use safer margins.

Reply:
@EdDev @oshoval I think that the scheduling latency of GO routines is non deterministic in this case because the underlying OS is of preemptive nature. Therefore I would like to suggest to set the `timeout` value to be very high (500msec. or even 1sec). I just don't want this unit test to validate whether the scheduling latency of the system is sane or insane.